vmm/src/vmm_config/drive.rs

```diff
@@ -158,72 +157,62 @@ impl BlockDeviceConfigs {
     }
 
     fn create(&mut self, block_device_config: BlockDeviceConfig) -> Result<()> {
-        // check if the path exists
         if !block_device_config.path_on_host.exists() {
-            return Err(DriveError::InvalidBlockDevicePath);
-        }
-
-        if self
+            Err(DriveError::InvalidBlockDevicePath)
+        } else if self
             .get_index_of_drive_path(&block_device_config.path_on_host)
             .is_some()
         {
-            return Err(DriveError::BlockDevicePathAlreadyExists);
-        }
-
-        // check whether the Device Config belongs to a root device
-        // we need to satisfy the condition by which a VMM can only have on root device
-        if block_device_config.is_root_device {
-            if self.has_root_block {
-                return Err(DriveError::RootBlockDeviceAlreadyAdded);
-            } else {
-                self.has_root_block = true;
-                self.read_only_root = block_device_config.is_read_only;
-                self.has_partuuid_root = block_device_config.partuuid.is_some();
-                // Root Device should be the first in the list whether or not PARTUUID is specified
-                // in order to avoid bugs in case of switching from partuuid boot scenarios to
-                // /dev/vda boot type.
-                self.config_list.push_front(block_device_config);
-            }
-        } else {
+            Err(DriveError::BlockDevicePathAlreadyExists)
+        } else if !block_device_config.is_root_device {
             self.config_list.push_back(block_device_config);
+            Ok(())
+        } else if self.has_root_block {
+            Err(DriveError::RootBlockDeviceAlreadyAdded)
+        } else {
+            self.has_root_block = true;
+            self.read_only_root = block_device_config.is_read_only;
+            self.has_partuuid_root = block_device_config.partuuid.is_some();
+            // Root Device should be the first in the list whether or not PARTUUID is specified
+            // in order to avoid bugs in case of switching from partuuid boot scenarios to
+            // /dev/vda boot type.
+            self.config_list.push_front(block_device_config);
+            Ok(())
         }
-
-        Ok(())
     }
 
     /// Updates a Block Device Config. The update fails if it would result in two
     /// root block devices.
     fn update(&mut self, mut index: usize, new_config: BlockDeviceConfig) -> Result<()> {
-        // Check if the path exists
         if !new_config.path_on_host.exists() {
-            return Err(DriveError::InvalidBlockDevicePath);
-        }
-
-        // Check if the root block device is being updated.
-        if self.config_list[index].is_root_device {
+            Err(DriveError::InvalidBlockDevicePath)
+        } else if self.config_list[index].is_root_device {
```

Comment:
Not sure this piece of refactoring improves readability.

I personally think that the original version was clearer in purpose and logic. Especially since the new version lost some valuable context comments.

Reply:
Okay, I'm reverting it to the original version